src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs

```diff
@@ -111,53 +116,119 @@ public virtual bool ShouldFormatSnippet(SnippetInfo snippetInfo)
 
         private void PopulateSnippetCaches()
         {
-            var updatedSnippets = GetSnippetInfoList();
-            var updatedSnippetShortcuts = GetShortcutsHashFromSnippets(updatedSnippets);
+            Debug.Assert(_expansionManager != null);
 
-            lock (cacheGuard)
-            {
-                snippets = updatedSnippets;
-                snippetShortcuts = updatedSnippetShortcuts;
-            }
-        }
+            // Ideally we'd fork execution here based on whether the expansion manager is an
+            // IExpansionManager or not. Unfortunately, we cannot mention that type by name until
+            // the Roslyn build machines are upgraded to Visual Studio 2015 Update 1. We therefore
+            // need to try using IExpansionManager dynamically, from a background thread. If that
+            // fails, then we come back to the UI thread for using IVsExpansionManager instead.
 
-        protected static HashSet<string> GetShortcutsHashFromSnippets(IList<SnippetInfo> updatedSnippets)
-        {
-            return new HashSet<string>(updatedSnippets.Select(s => s.Shortcut), StringComparer.OrdinalIgnoreCase);
+            var token = _waiter.BeginAsyncOperation(GetType().Name + ".Start");
+
+            Task.Factory.StartNew(async () => await PopulateSnippetCacheOnBackgroundWithForegroundFallback().ConfigureAwait(false),
+                            CancellationToken.None,
+                            TaskCreationOptions.None,
+                            TaskScheduler.Default).CompletesAsyncOperation(token);
         }
 
-        private IList<SnippetInfo> GetSnippetInfoList()
+        private async Task PopulateSnippetCacheOnBackgroundWithForegroundFallback()
         {
-            IVsExpansionEnumeration expansionEnumerator;
-            if (TryGetVsSnippets(out expansionEnumerator))
+            AssertIsBackground();
+
+            try
             {
-                return ExtractSnippetInfo(expansionEnumerator);
+                IVsExpansionEnumeration expansionEnumerator = await ((dynamic)_expansionManager).EnumerateExpansionsAsync(
```

Comment:
We have verified this is a managed -> managed call and isn't marshalling to the COM STA by accident?